roles/v2/pool/src/lib/job_negotiator/mod.rs

```diff
@@ -0,0 +1,118 @@
+            }
+        });
+    }
+}
+
+pub struct JobNegotiator {
+    downstreams: Vec<Arc<Mutex<JobNegotiatorDownstream>>>,
+    id: Id,
+}
+
+impl JobNegotiator {
+    pub async fn start(config: Configuration) {
+        let self_ = Arc::new(Mutex::new(Self {
+            downstreams: Vec::new(),
+            id: Id::new(),
+        }));
+        info!("JN INITIALIZED");
+        Self::accept_incoming_connection(self_, config).await;
+    }
+    async fn accept_incoming_connection(self_: Arc<Mutex<JobNegotiator>>, config: Configuration) {
+        let listner = TcpListener::bind(&config.listen_jn_address).await.unwrap();
+        while let Ok((stream, _)) = listner.accept().await {
+            let responder = Responder::from_authority_kp(
+                config.authority_public_key.clone().into_inner().as_bytes(),
+                config.authority_secret_key.clone().into_inner().as_bytes(),
+                std::time::Duration::from_secs(config.cert_validity_sec),
+            )
+            .unwrap();
+
+            let (receiver, sender): (Receiver<EitherFrame>, Sender<EitherFrame>) =
+                Connection::new(stream, HandshakeRole::Responder(responder)).await;
+            let setup_message_from_proxy_jn = receiver.recv().await.unwrap();
+            info!(
+                "Setup connection message from proxy: {:?}",
+                setup_message_from_proxy_jn
+            );
+
+            let setup_connection_success_to_proxy = SetupConnectionSuccess {
+                used_version: 2,
+                flags: 0b_0000_0000_0000_0000_0000_0000_0000_0001,
+            };
+            let sv2_frame: StdFrame =
+                PoolMessages::Common(setup_connection_success_to_proxy.into())
+                    .try_into()
+                    .unwrap();
+            let sv2_frame = sv2_frame.into();
+
+            info!("Sending success message for proxy");
+            sender.send(sv2_frame).await.unwrap();
+
+            let jndownstream = Arc::new(Mutex::new(JobNegotiatorDownstream::new(
+                receiver.clone(),
+                sender.clone(),
+            )));
+
+            self_
+                .safe_lock(|job_negotiator| job_negotiator.downstreams.push(jndownstream.clone()))
+                .unwrap();
+
+            println!(
+                "NUMBER of proxies JN: {:?}",
+                self_
+                    .safe_lock(|job_negotiator| job_negotiator.downstreams.len())
+                    .unwrap()
+            );
+            let coinbase_add_size = JobNegotiation::SetCoinbase(SetCoinbase {
+                coinbase_output_max_additional_size: crate::COINBASE_ADD_SZIE,
+                token: self_.safe_lock(|s| s.id.next()).unwrap() as u64,
+
+                coinbase_tx_prefix: crate::COINBASE_PREFIX.try_into().unwrap(),
+                coinbase_tx_suffix: crate::COINBASE_SUFFIX.try_into().unwrap(),
+            });
+            JobNegotiatorDownstream::send(jndownstream.clone(), coinbase_add_size)
```

Comment:
Is it intentional that this is just getting sent into the ether? jndownstream stay_alive loop won't do anything with this message, was that intentional?

Reply:
Yes, this is important: pool is sending this coinbase_add_size to the proxy and is the data needed for the JN to ask the TP for the new template and new prev hash. Once this data is sent to the JN, this is kept alive because this can be modified by pool and sent again, also the proxy is not expecting a closed connection with the upstram of the JN.